content/blog/2020-10-28-type-based-alias-analysis.md

```diff
@@ -0,0 +1,138 @@
+table that tracks the equivalence classes for a given type (such that for a 
+type `T` with entry `{T1, T2, …}`, an access path of type `T` may be a 
+reference to any `T1, T2, …`). Upon encountering an assignment from type `T1` 
+to a pointer of type `T2`, it adds `T1`’s entry to `T2`’s equivalence class. 
+Notably, this is an asymmetric operation; the assignment does not change `T1`’s 
+equivalence class. This is a more refined version of _TypeDecl_ above, so by 
+replacing _TypeDecl_ with lookups from this table (called _SMTypeRefs_ in the 
+paper, for “Selectively Merge Type References”), we have the final algorithm 
+_SMFieldTypeRefs_. 
+
+## Evaluation
+The paper does a surprisingly thorough evaluation job considering the paper 
+was written in 1998. The authors observe that previous alias analyses were 
+evaluated by their static properties, like the size of the set of alias pairs, 
+and their dynamic properties, such as the impact of an alias analysis on a 
+compiler optimization that uses it. The static evaluation, performed over a 
+selection of 10 benchmarks, finds that _TypeDecl_ performs much worse than 
+_FieldTypeDecl_ in terms of alias set size, but _SMFieldTypeRefs_ barely 
+improves on _FieldTypeDecl_ at all. It also shows that applying type-based 
+alias analysis interprocedurally generates huge numbers of aliases, rendering 
+this type of analysis infeasible for global optimization. 
+
+The authors then proceed to measure the impact of the three different analyses 
+on redundant load elimination, which moves invariant memory references out of 
+loops. They found that the number of redundant loads removed did increase with 
+the power of the analysis and that the amount of optimization improvement 
+between analyses was correlated with how much smaller the alias set became in 
+the static evaluation. On the other hand, when they considered the execution 
+time post-redundant-load-elimination, they found that all three analyses 
+performed similarly, averaging a 4% speedup from RLE. Thus, perhaps 
+counterintuitively, more precision in the alias analysis doesn’t necessarily 
+yield significant gains in runtime speed.
+
+However, the authors note that static properties don’t directly correlate with 
+performance benefits, and don’t lend themselves to comparing two different 
+alias analyses – in both cases, it’s hard to tell whether the disambiguated 
+pointers will be relevant for the intended use case. Dynamic properties, on the 
+other hand, suffer from overspecificity – it’s difficult to evaluate general 
+efficacy of an alias analysis from performance improvements on a few specific 
+optimizations and benchmarks. Worst of all, neither evaluation can give an idea 
+of how much better an analysis could be.
+
+To alleviate these problems, the paper introduces the concept of limit 
+evaluation to figure out whether there were missed optimization opportunities 

```

Comment:
Maybe italicize "limit evaluation"? Also, FWIW, I think the more common term for this kind of thing today is "limit study."

Reply:
done! i kept "limit evaluation" because that's how they introduced it in the paper; notably, it's impossible to google for the term "limit study" without getting a bunch of math stuff 😛